File: InterviewCakeProblems2/permutation_palindrome.py
def permutation_palindrome(input_str):
    unpaired_chars = set()

    for char in input_str:
        if char in unpaired_chars:
            unpaired_chars.remove(char)
        else:
            unpaired_chars.add(char)
    return len(unpaired_chars) <= 1
    
print (permutation_palindrome("civic")) ## should return true
print (permutation_palindrome("ivicc")) ## should return true
print (permutation_palindrome("civil")) ## should return false
print (permutation_palindrome("livci")) ## should return false

File: InterviewCakeProblems2/permutation_palindrome.rb
require 'set';

def permutation_palindrome(input_str)
    unpaired_chars = Set.new()

    input_str.each_char do |current_char|
        if unpaired_chars.include?(current_char)
            unpaired_chars.delete(current_char)
        else
            unpaired_chars.add(current_char)
        end
    end
    unpaired_chars.size <= 1
end 

p permutation_palindrome("civic"); ##should return true
p permutation_palindrome("ivicc"); ##should return true
p permutation_palindrome("civil"); ##should return false
p permutation_palindrome("livci"); ##should return false